spark/src/test/java/org/apache/iceberg/spark/data/TestSparkOrcUnions.java

```diff
@@ -0,0 +1,108 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.spark.data;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.iceberg.Files;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.orc.ORC;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Types;
+import org.apache.orc.OrcFile;
+import org.apache.orc.TypeDescription;
+import org.apache.orc.Writer;
+import org.apache.orc.storage.ql.exec.vector.BytesColumnVector;
+import org.apache.orc.storage.ql.exec.vector.LongColumnVector;
+import org.apache.orc.storage.ql.exec.vector.UnionColumnVector;
+import org.apache.orc.storage.ql.exec.vector.VectorizedRowBatch;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+
+public class TestSparkOrcUnions {
+  private static final int BATCH_SIZE = 50;
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  @Test
+  public void testComplexUnion() throws IOException {
+    TypeDescription orcSchema =
+        TypeDescription.fromString("struct<unionCol:uniontype<int,string>>");
+
+    Schema expectedSchema = new Schema(
+        Types.NestedField.optional(0, "unionCol", Types.StructType.of(
+            Types.NestedField.optional(1, "tag_0", Types.IntegerType.get()),
+            Types.NestedField.optional(2, "tag_1", Types.StringType.get())))
+    );
+
+    Configuration conf = new Configuration();
+
+    temp.delete();
+    String basePath = temp.getRoot().getAbsolutePath();
+    Path path = new Path(basePath + "/test.orc");
+
+    Writer writer = OrcFile.createWriter(path,
+        OrcFile.writerOptions(conf)
+            .setSchema(orcSchema));
+
+    VectorizedRowBatch batch = orcSchema.createRowBatch();
+    LongColumnVector longColumnVector = new LongColumnVector(BATCH_SIZE);
+    BytesColumnVector bytesColumnVector = new BytesColumnVector(BATCH_SIZE);
+    UnionColumnVector complexUnion = new UnionColumnVector(BATCH_SIZE, longColumnVector, bytesColumnVector);
```

Comment:
I see you are manually creating 3 new `ColumnVector`s, but actually, the `VectorizedRowBatch` built from the orc schema has already the `UnionColumnVector` (`batch.cols[0]`) and also the internal primitive column vectors created for you. But this is just a nit, it's up to you if you want to change it or not.